### PR TITLE
Feature/di refactoring

### PR DESCRIPTION
**Pull request by roxblnfk**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️

`migrate/generate` now shows changes (added `ShowChangesGenerator`)
Added typical mappers: `SoftDeletedMapper` and `TimestampedMapper`
`CycleOrmHelper` has been refactored

### CODE REVIEW

File: src/Generator/ShowChangesGenerator.php [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    public function __construct(OutputInterface $output)
+    {
+        $this->output = $output;
+    }
+
+    public function run(Registry $registry): Registry
+    {
+        $this->output->writeln('<info>Detecting schema changes:</info>');
+        $this->changes = [];
+        foreach ($registry->getIterator() as $e) {
```
Comment (samdark) [inline src/Generator/ShowChangesGenerator.php]: What is `$e` in this case?

File: src/Mapper/SoftDeletedMapper.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+use Cycle\ORM\Command\CommandInterface;
+use Cycle\ORM\Command\Database\Update;
+use Cycle\ORM\Context\ConsumerInterface;
+use Cycle\ORM\Heap\Node;
+use Cycle\ORM\Heap\State;
+use Cycle\ORM\Mapper\Mapper;
+
+/**
+ * @Table(
+ *      columns={"deleted_at": @Column(type="datetime", nullable=true)}
```
Comment (samdark) [inline src/Mapper/SoftDeletedMapper.php]: Some additional documentation on usage would be very helpful.

File: src/Mapper/SoftDeletedMapper.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ */
+class SoftDeletedMapper extends Mapper
+{
+    public function queueDelete($entity, Node $node, State $state): CommandInterface
+    {
+        // identify entity as being "deleted"
+        $state->setStatus(Node::SCHEDULED_DELETE);
+        $state->decClaim();
+
+        $cmd = new Update(
```
Comment (samdark) [inline src/Mapper/SoftDeletedMapper.php]:
```suggestion
        $command = new Update(
```

File: src/Mapper/TimestampedMapper.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * @Table(
+ *      columns={"created_at": @Column(type="datetime"), "updated_at": @Column(type="datetime")}
+ * )
+ */
+class TimestampedMapper extends Mapper
+{
+    public function queueCreate($entity, Node $node, State $state): ContextCarrierInterface
+    {
+        $cmd = parent::queueCreate($entity, $node, $state);
```
Comment (samdark) [inline src/Mapper/TimestampedMapper.php]:
```suggestion
        $command = parent::queueCreate($entity, $node, $state);
```

File: src/Exception/BadDeclarationException.php [hunk 5]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace Yiisoft\Yii\Cycle\Exception;
+
+use Exception;
+
+class BadDeclarationException extends Exception
```
[the remark this answers is not in the document]
Comment (roxblnfk) [reply src/Exception/BadDeclarationException.php]: I thought about it and did not dare to add a new dependency

File: src/SchemaConveyorInterface.php [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    public const STAGE_INDEX = 'index';
+    public const STAGE_RENDER = 'render';
+    public const STAGE_USERLAND = 'userland';
+    public const STAGE_POSTPROCESS = 'postprocess';
+
+    /**
+     * @param string $stage
+     * @param mixed  $generator
+     */
+    public function addGenerator(string $stage, $generator): void;
```
[the remark this answers is not in the document]
Comment (samdark) [reply src/SchemaConveyorInterface.php]: OK.

File: config/common.php [hunk 7]
```diff
@@ -15,12 +17,11 @@
...
-    CycleOrmHelper::class => [
-        '__class' => CycleOrmHelper::class,
-        'addEntityPaths()' => [
-            'paths' => $params['cycle.common']['entityPaths'],
-        ],
-    ],
+
+    SchemaConveyorInterface::class => static function (ContainerInterface $container) use (&$params) {
+        $conveyor = new AnnotatedSchemaConveyor($container);
+        $conveyor->addEntityPaths($params['cycle.common']['entityPaths']);
```
[the remark this answers is not in the document]
Comment (roxblnfk) [reply config/common.php]: I'll postpone this change for later so as not to break BC

File: src/Generator/AnnotatedSchemaConveyor.php [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+    public function getTableNaming(): int
+    {
+        return $this->tableNaming;
+    }
+
+    /**
+     * @param string|string[] $paths
+     */
+    public function addEntityPaths($paths): void
```
[the remark this answers is not in the document]
Comment (roxblnfk) [reply src/Generator/AnnotatedSchemaConveyor.php]: accepted array of paths only

File: src/Mapper/SoftDeletedMapper.php (hunk 2 above)
Comment (roxblnfk) [reply src/Mapper/SoftDeletedMapper.php]: Added `@See` block and short comment

File: src/Generator/AnnotatedSchemaConveyor.php [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+namespace Yiisoft\Yii\Cycle\Generator;
+
+use Cycle\Annotated;
+use Doctrine\Common\Annotations\AnnotationRegistry;
+use Psr\Container\ContainerInterface;
+use Spiral\Tokenizer\ClassLocator;
+use Symfony\Component\Finder\Finder;
+use Yiisoft\Aliases\Aliases;
+
+class AnnotatedSchemaConveyor extends SchemaConveyor
```
[the remark this answers is not in the document]
Comment (roxblnfk) [reply src/Generator/AnnotatedSchemaConveyor.php]: `AnnotatedSchemaConveyor` marked as Final

**Comment by samdark** [comment]
Merged. Thank you!

**Pull request merged**